### PR TITLE
Fixed route editing and drawing mask

### PR DESCRIPTION
**Pull request by vladyslav-tk**
- Fixed editing of route segments and control points
- Fixed drawing mask on high DPI devices

### CODE REVIEW

File: src/interaction/TrackManager.ts [hunk 1]
```diff
@@ -651,28 +651,42 @@ export default class TrackManager<POIMeta> {
...
 
-    const coordinates = [
-      [extent[0], extent[1]], // Bottom-left
-      [extent[0], extent[3]], // Top-left
-      [extent[2], extent[3]], // Top-right
-      [extent[2], extent[1]], // Bottom-right
-    ];
+    const pixelRatio = (canvas.width / viewportWidth) * frameState.pixelRatio;
+    console.assert(
+        pixelRatio === (canvas.height / viewportHeight) * frameState.pixelRatio,
```
Comment (gberaudo) [inline src/interaction/TrackManager.ts:666]: Can we call it canvasViewportRatio?

File: src/interaction/TrackUpdater.ts [hunk 2]
```diff
@@ -73,10 +73,11 @@ export default class TrackUpdater {
       if (after) {
         geometryUpdates.push(this.router.snapSegment(after, modifiedControlPoint, pointTo));
       }
+      const results = await Promise.all(geometryUpdates);
```
Comment (gberaudo) [inline src/interaction/TrackUpdater.ts:76]: This is done first so that it can throw without side effect?
Can you add a comment?

**Review by gberaudo — APPROVED** [review]

File: src/interaction/TrackManager.ts [hunk 3]
```diff
@@ -651,28 +651,42 @@ export default class TrackManager<POIMeta> {
     if (!extent?.length) return;
     const viewport = event.target.getViewport();
     const canvases = viewport.getElementsByTagName('canvas');
+    const frameState = event.frameState;
+    const viewportWidth = Number(
+        (frameState.size[0] * frameState.pixelRatio).toFixed(),
+    );
+    const viewportHeight = Number(
+        (frameState.size[1] * frameState.pixelRatio).toFixed(),
+    );
```
[the remark this answers is not in the document]
Comment (vladyslav-tk) [reply src/interaction/TrackManager.ts]: no, we can use Math.round, I copied it from SM print

File: src/interaction/TrackManager.ts (hunk 1 above)
Comment (vladyslav-tk) [reply src/interaction/TrackManager.ts]: done

File: src/interaction/TrackUpdater.ts (hunk 2 above)
Comment (vladyslav-tk) [reply src/interaction/TrackUpdater.ts:77]: we should wait until snapping is done before equalize coordinates, I added a comment

**Pull request merged**